rust/src/sysroot_upgrade.rs

```diff
@@ -3,48 +3,82 @@
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 
 use crate::cxxrsutil::*;
-use crate::ffi::ContainerImport;
-use std::convert::TryInto;
+use crate::ffi::{output_message, ContainerImport};
+use anyhow::Result;
+use ostree_container::store::LayeredImageImporter;
+use ostree_container::store::PrepareResult;
+use ostree_container::OstreeImageReference;
+use ostree_ext::container as ostree_container;
+use ostree_ext::ostree;
+use std::convert::TryFrom;
 use std::pin::Pin;
 use tokio::runtime::Handle;
 
+async fn pull_container_async(
+    repo: &ostree::Repo,
+    imgref: &OstreeImageReference,
+    current_digest: Option<&str>,
+) -> Result<ContainerImport> {
+    let unchanged = || ContainerImport {
+        changed: false,
+        ostree_commit: "".to_string(),
+        image_digest: "".to_string(),
+    };
+    output_message(&format!("Pulling manifest: {}", &imgref));
+    match &imgref.sigverify {
+        ostree_container::SignatureSource::OstreeRemote(_) => {
```

Comment:
The logic here is that right now if we're pulling an `ostree-remote-image` then since we don't support layered, ostree-signed imports, we default to un-encapsulation.  If we're pulling a not-ostree-signed image, then we use layering.

But basically this is https://github.com/ostreedev/ostree-rs-ext/pull/99#issuecomment-926677281